packages/mui-styled-engine/src/cx.js

```diff
@@ -0,0 +1,149 @@
+/* eslint-disable import/prefer-default-export */
+/* eslint-disable @typescript-eslint/naming-convention */
+/* eslint-disable import/first */
+/* eslint-disable @typescript-eslint/no-use-before-define */
+/* eslint-disable no-labels */
+/* eslint-disable no-restricted-syntax */
+/* eslint-disable @typescript-eslint/no-shadow */
+import { useMemo } from 'react';
+import { serializeStyles } from '@emotion/serialize';
+import { insertStyles, getRegisteredStyles } from '@emotion/utils';
+import { __unsafe_useEmotionCache as useEmotionCache } from '@emotion/react';
+import createCache from '@emotion/cache';
+import { classnames } from './tools/classnames';
+
+const { createCx } = (() => {
+  function merge(registered, css, className) {
+    const registeredStyles = [];
+
+    const rawClassName = getRegisteredStyles(registered, registeredStyles, className);
+
+    if (registeredStyles.length < 2) {
+      return className;
+    }
+
+    return rawClassName + css(registeredStyles);
+  }
+
+  function createCx(params) {
+    const { cache } = params;
+
+    const css = (...args) => {
+      const serialized = serializeStyles(args, cache.registered);
+      insertStyles(cache, serialized, false);
+      const className = `${cache.key}-${serialized.name}`;
+
+      scope: {
+        const arg = args[0];
+
+        if (!matchCSSObject(arg)) {
+          break scope;
+        }
+
+        increaseSpecificityToTakePrecedenceOverMediaQueries.saveClassNameCSSObjectMapping(
+          cache,
+          className,
+          arg,
+        );
+      }
+
+      return className;
+    };
+
+    const cx = (...args) => {
+      const className = classnames(args);
+
+      const feat27FixedClassnames =
+        increaseSpecificityToTakePrecedenceOverMediaQueries.fixClassName(cache, className, css);
+
+      return merge(cache.registered, css, feat27FixedClassnames);
+    };
+
+    return { cx };
+  }
+
+  return { createCx };
+})();
+
+/** Will pickup the contextual cache if any */
+export function useCx() {
+  const cache = useEmotionCache();
+
+  const { cx } = useMemo(
+    () => createCx({ cache: cache ?? createCache({ key: 'never' }) }),
+    [cache],
+  );
+
+  return { cx };
+}
+
+// https://github.com/garronej/tss-react/issues/27
+const increaseSpecificityToTakePrecedenceOverMediaQueries = (() => {
```

Comment:
There is no 'bug' strictly speaking in the default `cx`, it works as the specs says it work; but not as the users expect it to work.  
Whether or not the increase of specificity to override media queries should be the default behavior of `cx` is up to debate.  
For TSS I had no choice but to implement the feature. Users were able to override media queries in v4 so I had to implement it or it would have made it very hard for some projects to upgrade to MUIv5.  

I personally think that having this increase of specificity is a hard requirement for us.  
If I have a button that, by default, is red for large screens and blue otherwise and I decide to make it white by passing a custom class. I expect it to be white regardless of the screen size.    

If I kill this feature MUI and TSS will end up with an avalanche of issues like "BUG: My custom style only applies on small screen size".

Reply:
> Interesting, I've tried replicating this with pure CSS, and looks like it works as you are suggesting (if we assume the order of the classes in the CSS file I have would correspond with the order of the classes added in the cx util): https://codepen.io/mnajdova/pen/VwyBEyX

Yeah, this is how it already works in [Emotion](https://codesandbox.io/s/amazing-elbakyan-x19ikj?file=/src/App.js) because we just "flatten" everything into a single rule~. Media query doesn't increase the specificity anyhow so it "just works". However, at times this might be confusing because sometimes people only want to override the "base" declaration without affecting the media query one and there is no easy way to do this.

This might depend on how exactly you merge styles here, if you nest them etc. I would appreciate a codesandbox or something of the "broken" behavior in MUI - then I could perhaps comment further because right now I don't fully understand it as the basic case should work exactly like that.

Reply:
Hi @Andarist, thank you for your involvement,  
I'll fix you a sandbox ASAP

Reply:
No worries, just ping me when it's ready

Reply:
@Andarist @mnajdova 
There you go: https://stackblitz.com/edit/tss-react-frmsvm?file=index.tsx  
On this sandbox you have, side by side, the vanilla `cx` and the patched `cx`'s behaviors.  
@mnajdova **Thank you** for coming up with this sandbox! Now I understand why it was working in v4.  
The fact that it is working like that with raw CSS is a strong argument in favor of classifying the current behavior of `cx` as "bug".  

I would much prefer having a more legit implementation of this behavior rather than my ad hock solution.

Reply:
Ah, I've prepared my codesandbox all wrong 🤣 and it got me to the wrong conclusions, see [the fixed one](https://codesandbox.io/s/amazing-elbakyan-x19ikj?file=/src/App.js). 

So ye - the problem here is somewhat the opposite one. Note that while this doesn't exactly work like in raw CSS it works like in some well-established CSS preprocessors. Take for example this SCSS input:
```scss
[namespace] {
    background-color: hotpink;
    
    @media screen and (min-width: 1px) {
        background-color: red;
    }

    background-color: lightgreen;
}
```
and its output
```css
[namespace] {
  background-color: hotpink;
  background-color: lightgreen;
}
@media screen and (min-width: 1px) {
  [namespace] {
    background-color: red;
  }
}
```

The problem is that each nesting level has an "implicit" rule slot that gets automatically created there and everything from within that context gets merged together. This is why the declaration that is after the `@media` gets merged with the one that is before it. This is also how mixins etc work in SCSS (and probably in Less and Stylus too).

So while you could fix the "raw CSS" expectations here... you most likely would break the other group of people. It feels to me that most CSS-in-JS solutions are closer to CSS preprocessors than to raw CSS and thus this result makes more sense.

Also, note that the `patchedCss` example from the codesandbox above is prone to injection order because you have 2 class names on that element. So it only works "accidentally". And this becomes even less reliable with stuff like Suspense, streaming SSR etc - with those your CSS just has to be idempotent on injection order, otherwise, you risk very hard to debug (and random(ish)!) visual bugs.

Reply:
> Also, note that the patchedCss example from the codesandbox above is prone to injection order because you have 2 class names on that element. So it only works "accidentally".  

You are right, sorry about that my sandbox was totally working by coincidence.  
This is what I should have submitted: https://stackblitz.com/edit/tss-react-t9hqtn?file=index.tsx  

> And this becomes even less reliable with stuff like Suspense, streaming SSR etc  

And it breaks [when nested `cx` are used...](https://stackblitz.com/edit/tss-react-bmfwhm?file=index.tsx)...  
My hack only works in the most basic use case, it seems to do it for peoples, I had no complains about it, but I don't like it either. 

@mnajdova it was dumb of me to get my hack in the MUI codebase. It only works if both the patched `css` and the patched `cx` are used. The change would have to be made upstream in `@emotion` to make sense.
I will delete it but we still need to address the issue. 

> So while you could fix the "raw CSS" expectations here... you most likely would break the other group of people. It feels to me that most CSS-in-JS solutions are closer to CSS preprocessors than to raw CSS and thus this result makes more sense.  

I get your point, but understand our situation here. A MUI user do:  
```tsx
<SomeMuiComponent 
    classes={{
        "root": css({ backgroundColor: "white" })
    }}
/>
```  
He expects the background color of the component to be white.  

What are your thoughts?

Reply:
> This is what I should have submitted: https://stackblitz.com/edit/tss-react-t9hqtn?file=index.tsx

This looks like a different problem though. All of the styles are contained in a single class (`.css-ol8gfl-MuiButtonBase-root-MuiButton-root`) and we can see how `color: #1976d2` just appears after your `color: pink` there. This indicates that your `classes.root` are not simply appended to the styles defined in the `Button` but rather somewhere in the middle of those styles.

Reply:
It all depends on the established patterns for stylish, composition, etc. How much of this information is embedded in the components, how low-level they and how the consumer can customize the behavior.

I, personally, remember that I was surprised that `@media` query doesn't actually increase the specificity when I first learned about it. After all - conceptually it is providing more information, it is more specific.

Maybe this is more of a documentation+helper problem and not truly about the styling engine? From what I understand - this is a behavior~ change from v4 here. Do you have a codesandbox that would showcase how this used to work?

Reply:
@Andarist Yes it could work, this PR is about making it work.  
If it isn't currently working it's because, MUI is using `clsx` instead of `cx`.  
If you'd run this example with my fork, you'd get a pink text as expected.  

We might face issues later on related to the fact that emotion grant higher specificity to rules within media query but let's forget about that for now. We'll see if it *actually* ends up being a problem or not.

Reply:
Yes, I know that this PR is trying to make it work - but I'm not completely sold on the whole `clsx` vs `cx` difference here.

If we take a look at the generated class name:
<img width="445" alt="Screenshot 2022-05-09 at 17 23 28" src="https://user-images.githubusercontent.com/9800850/167443136-a69ec3c4-3e82-4786-aa0a-f9ecdf471eb4.png">

then we might notice that the `color: pink;` is here - it's just not last. So Emotion classes can be "merged" just fine, it's just that they are merged in a different order.

I might be wrong but it feels like this could work with `clsx`, if only other changes would be introduced to the codebase.

Reply:
I mean, I'd really *love to be proven wrong* but I am pretty sure there is no other way to make it work but to use `cx`.

Reply:
Thank you for taking the time to prove your point.  
I was stuck on the idea that:  
```ts
clsx(classes.root, className) === `${classes.root} ${className}`
```
which I assumed to be equivalent to 
```ts
`${classname} ${classes.root}`
```  
I thought only `cx` was able to enforce priority of one class over another.  

Turns out I was wrong, I am not familiar enough with the inner working of emotion.  

It’s good news.

Reply:
What u are saying is correct but in this case the resulting `className` is passed to Emotion and it can still handle decoding that

Reply:
Ok thx, make sense!